### PR TITLE
Fix detection of loop headers

### LINKED ISSUE #164: Could not find loop header
Early returns in while-loops, as in:
```
fn example() -> bool {
    let mut j = 0;
    let loop_len = 5;
    #[invariant(loop_bound2, 0usize <= j && j <= loop_len)]
    while 0 < 5 {
        return true;
    }
    return true;
}
```
Gives the following error `thread 'rustc' panicked at 'Could not find loop header', creusot/src/gather_invariants.rs:146:17`

### PR DESCRIPTION
**Pull request by Lysxia**
Fixes #164 

We desugar all loops using `loop`, we put a noop marker right before the `loop` keywords, and the invariants right after. We search for the loop header backwards from the invariants and stop when we hit a marker, in which case the invariants become assertions.

In hindsight this should work just as well with the invariants before `loop` and the marker after it, but the implementation turns out to be slightly more convenient this way because it lets us search for the marker block by block instead of statement by statement.

### CODE REVIEW

**Comment by jhjourdan** [comment]
What is the plan for handling `for` loop? In the case of `for` loops, the environment can be different from the environment before the loop (and hence shadow variables), and some mutation happen just before the place where the invariant is inserted. But still, the invariant should be interpreted in the outer scope.

File: creusot-contracts-proc/src/invariant.rs [hunk 1]
```diff
@@ -172,10 +180,10 @@ fn desugar_for(mut invariants: Vec<Invariant>, f: ExprForLoop) -> TokenStream {
         let mut #it = ::std::iter::IntoIterator::into_iter(#iter);
         let #iter_old = snapshot! { #it };
         let mut #produced = snapshot! { ::creusot_contracts::logic::Seq::EMPTY };
-        #(#invariants)*
         #(#outer)*
         #lbl
         loop {
+            #(#invariants)*
```
[the remark this answers is not in the document]
Comment (Lysxia) [reply creusot-contracts-proc/src/invariant.rs:186]: `x` is bound inside the `match` below so it's not in scope at the top of `loop`.

[the remark this answers is not in the document]
Comment (Lysxia) [reply creusot-contracts-proc/src/invariant.rs:186]: OK This one is a problem. I was trying to wrap the RHS of `while let` as `{ #(#invariants)* #rhs }` but introducing a block changes the semantics (some variables are moved).

I'm going back to the original approach and try to fix the heuristic to detect loops.

**Comment by jhjourdan** [comment]
> I had to change the `invariant_moves` test case because the curly braces in the RHS of `while let Some(_) = { ... }` are not allowed in `let .... else ...`...

That's not very satisfying. Why isn't this syntax allowed? What if we add a pair of parentheses ?

Why didn't you use `if let ... { ... } else { break }` instead?

**Comment by Lysxia** [comment]
`if let` works better indeed!

**Comment by jhjourdan** [comment]
> `if let` works better indeed!

Great!

It's weird, though, that "let else" does not support this.

What's happening if the pattern is irrefutable? Does rustc complain about dead code?

**Pull request merged**